pkg/agent/nodeportlocal/portcache/port_table_windows.go

```diff
@@ -0,0 +1,273 @@
+//go:build windows
+// +build windows
+
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package portcache
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/nodeportlocal/rules"
+)
+
+func addRuleForPort(podPortRules rules.PodPortRules, port int, podIP string, podPort int, protocol string) ([]ProtocolSocketData, error) {
+	// port needs to be available for all supported protocols: we want to use the same port
+	// number for all protocols and we don't know at this point which protocols are needed.
+	protocols := make([]ProtocolSocketData, 0, len(supportedProtocols))
+	err := podPortRules.AddRule(port, podIP, podPort, protocol)
```

Comment:
TODO

Reply:
Comment updated.
@antoninbas I tried to build this scenario on my test environment. I created a service which supports both UDP and TCP on the same port and enable the pods' NPL annotation feature. The result showed windows agent can insert both UDP and TCP NetNatStaticmapping rules for the same port. 
Here is the rule insertion result on my windows host:

StaticMappingID               : 129
NatName                       : antrea-nat
Protocol                      : TCP
RemoteExternalIPAddressPrefix : 0.0.0.0/0
ExternalIPAddress             : 0.0.0.0
ExternalPort                  : 61003
InternalIPAddress             : 192.168.3.217
InternalPort                  : 80
InternalRoutingDomainId       : {00000000-0000-0000-0000-000000000000}
Active                        : True

StaticMappingID               : 130
NatName                       : antrea-nat
Protocol                      : UDP
RemoteExternalIPAddressPrefix : 0.0.0.0/0
ExternalIPAddress             : 0.0.0.0
ExternalPort                  : 61003
InternalIPAddress             : 192.168.3.217
InternalPort                  : 80
InternalRoutingDomainId       : {00000000-0000-0000-0000-000000000000}
Active                        : True

Reply:
I want to clarify the scenario I am talking about:

1. A Pod is scheduled on that Node with exposes a TCP service (port 80) using NPL: this is the first time we need to allocate a port for NPL, so we pick the first in the range (61000).
2. A UDP server is started on the Node on port 61000.
3. The Pod (from the first step) is updated so that UDP port 80 is also exposed. We need to use the same host port number as for TCP (this is how NPL works currently), so we try using 61000, but the port is already taken.

In the Linux implementation, step 2 will fail (unless something like `SO_REUSEADDR` is used). What about the Windows implementation?

Reply:
I see.. I found I can still start a udp server on the same port after inserting tcp rule. It seems the socket bind function matches tuples that includes protocol? I thought inserting netnatstaticmapping rules with any protocol will reserving the port for all protocols but this result showed I was wrong. Do you have any ideas to avoid this mistake? @antoninbas

Reply:
I think there are a few options, but some of them requires changing the behavior of the NPL feature. Would this idea work: we always create NetNat mapping for TCP + UDP using the same port number, even if (at the moment at least) the Pod only cares about once protocol (e.g. TCP). If we fail to insert a mapping, we clean up and try the next port. _However_, the the Pod only care about one protocol (e.g. TCP), we use a "fake" `InternalIPAddress` for the other protocol (e.g.  reserved link local address). If the other protocol gets added later, we change the NetNat mapping accordingly to use the Pod's IP address. The disadvantage of this approach is that I think if someone tries to access the port using the "unsupported" protocol, the behavior is a bit different.

Reply:
Hi Antonin, I want to know why we must ensure the Node port for UDP and TCP uses the same number for a single Pod? What if we use a different Node port for UDP from TCP traffic on a Pod even though the ports exposed in Pod are the same?

Reply:
@wenyingd I agree that this is not ideal, but this is because the NPL annotation doesn't include the protocol at the moment:
```yaml
  annotations:
    nodeportlocal.antrea.io: '[{"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61002}]'
```
IIRC, this is because we originally only supported TCP (so the protocol wasn't included) and when we added support for UDP, it was quite easy to use the same host port on Linux, without the need to change the annotation. We didn't think of how it may not be convenient on Windows...

We can make a change to the annotation to include the protocol string if we don't have another choice, but that may be considered a non backwards-compatible change to a user-facing API. At the very least we would need to check with the AVI teams, since we know they are using this feature.

Reply:
Hi @antoninbas , Shuyang and  I saw the annotations yesterday and we found that "protocols" is supported in the annotation. But the point you mentioned is correct that we should consider the change to be backwards-compatible to a user-facing API. As it is the first release for Windows, and I think we don't need to assume the case that Windows Pods run on Linux Node or Linux Pods run on Windows Node. So I think maybe it is a candidate to have different behavior on Windows if AVI teams accept it, right?

As for your suggestion that using a fake Windows NetNatStaticMapping to take the port with a different protocol, the issue is we need to maintain a new Windows NetNat to create fake static-mappings, and remove them from the new netnat to add on the validate natnat. It can't be operated by a simple replace operation. And we need also to consider reconciling process on the two netnats.

Another thought is using socket to take the port with a unconfigured protocol instead of NetNat, and the  we should think of closing the socket before creating the real NetNatStaticMapping.

Actually, we also have a third thought that we could rollback to choose a new port for both protocols on the case that the 1st protocal is successful but 2nd is failed. But this also requires a modification on the existing annotation, which might introduce backward-compatible issues.

What do you think?

Reply:
I completely forgot that the annotation had a protocols list: https://github.com/antrea-io/antrea/blob/fe21260d84f66b1c89caf539a90e9abb61e6104f/pkg/agent/nodeportlocal/k8s/annotations.go#L45

I was going by the documentation, which apparently is not completely up-to-date: https://github.com/antrea-io/antrea/blob/main/docs/node-port-local.md

Ideally the annotation would really look like this:
```go
type NPLAnnotation struct {
	PodPort   int      `json:"podPort"`
	NodeIP    string   `json:"nodeIP"`
	NodePort  int      `json:"nodePort"`
	Protocol string `json:"protocol"`
}
```
With a different mapping entry per protocol.

I feel like we should do the right thing while the feature is still Beta, even if the change is a bit disruptive. We could update the annotation to this:
```go
type NPLAnnotation struct {
	PodPort   int      `json:"podPort"`
	NodeIP    string   `json:"nodeIP"`
	NodePort  int      `json:"nodePort"`
	Protocol string `json:"protocol"`
        Protocols []string `json:"protocols"` // deprecated, array with a single member which is equal to the Protocol field
}
```
Instead of publishing this annotation:
```
nodeportlocal.antrea.io: '[{"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61002,"protocols":["tcp","udp"]}]'
```
we would have:
```
nodeportlocal.antrea.io: '[
  {"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61002,"protocol":"tcp","protocols":["tcp"]},
  {"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61002,"protocol":"udp","protocols":["udp"]}
]'
```
or:
```
nodeportlocal.antrea.io: '[
  {"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61002,"protocol":"tcp","protocols":["tcp"]},
  {"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61009,"protocol":"udp","protocols":["udp"]}
]'
```
with a different port (both of the above annotations would be valid)

Keeping the `protocols` field may help with backwards compatibility based on the assumptions made by the consumers of these annotations.

Could you check with the AVI folks if that would be acceptable? Make sure to emphasize that we can now have 2 different mappings for the same (pod IP, pod port) with one for UDP and one for TCP, and that the node port may or may not be the same for both mappings.

If it's not acceptable we can opt for the other solution, which is to introduce a `v2` for the annotation. We would have a new annotation `nodeportlocal.v2.antrea.io` or `v2.nodeportlocal.antrea.io`. For Linux Nodes, we would publish both versions and for Windows Nodes, only the second version. For a Linux Node it would look like this:
```go
nodeportlocal.antrea.io: '[{"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61002,"protocols":["tcp","udp"]}]'
v2.nodeportlocal.antrea.io: '[
  {"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61002,"protocol":"tcp","protocols":["tcp"]},
  {"podPort":8080,"nodeIP":"10.10.10.10","nodePort":61002,"protocol":"udp","protocols":["udp"]} // same nodePort!
]'
```

Eventually we would remove the first version of the annotation (after 9 months to 1 year). I'd rather go with the first solution if possible. It will be simpler for us.

Reply:
BTW, let's not go with anything which requires possible rollbacks, it's too error prone.

Reply:
Thanks for all comments. 
I was trying to implement Antonin’s first idea in past two days.  But I found another problem when I created another NPL table to reserve port by using link local address:
NetNat rules on windows **only** allowed inserting two rules with same external port and different protocol when they had the same Natname, which is the netnat table here. Otherwise it failed and return “The object already exists” error message.
For example:
```
Add-NetNatStaticMapping  -NatName antrea-nat -ExternalIPAddress 0.0.0.0 -ExternalPort 61021 -InternalIPAddress 192.168.2.39 -Protocol TCP
Add-NetNatStaticMapping  -NatName antrea-nat -ExternalIPAddress 0.0.0.0 -ExternalPort 61021 -InternalIPAddress 192.168.2.39 -Protocol UDP
```
These two rules can be inserted successfully.

However, when I ran:
```
Add-NetNatStaticMapping  -NatName antrea-nat -ExternalIPAddress 0.0.0.0 -ExternalPort 61021 -InternalIPAddress 192.168.2.39 -Protocol TCP

Add-NetNatStaticMapping  -NatName antrea-nat-reserve -ExternalIPAddress 0.0.0.0 -ExternalPort 61021 -InternalIPAddress 169.254.0.1 -Protocol UDP
Error:
Add-NetNatStaticMapping : The object already exists.
At line:1 char:1
+ Add-NetNatStaticMapping  -NatName antrea-nat-reserve -ExternalIPAddre ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (MSFT_NetNatStaticMapping:root/StandardCi...atStaticMapping) [Add-NetNatStaticMapping], CimException
    + FullyQualifiedErrorId : Windows System Error 5010,Add-NetNatStaticMapping
```

The commands always failed if two NatNames are different.

Another way to solve this problem is to listen port for reserved protocol port, which can implement the same behavior compared to Linux. Here is the updated function details:
1. For getFreePort function: It needs to listen port if there is no real rules on its protocol. But we still keep using real netnat rule to reserve its protocol port at the same time.  This happens on getFreePort stage.
2. For next Addrule function: If it needs to insert new netnat rule to a stateopen port, the socket will be closed before inserting rules. 
3. For Deleterule function: It need to reopen socket after deletion to make sure the port is reserved until release.
4. ReleaseNPLPort function can delete all related netnat rules and close socket to release port.

Wenying and I agreed that Using two NPL tables and listening port are the two most pragmatic solution to solve current problem. Since the first solution had new issue, I tend to use the second one.

PR updated.
https://github.com/antrea-io/antrea/pull/3453/commits/fae623ac20f69a10ec8edff92249cad0a82380f0 

Do you have any suggestions? Thanks! @antoninbas  @wenyingd

Reply:
No a big fan of a solution like this, since there is always a (slim) possibility of someone else grabbing the port between the time when we close the socket and the time when we insert the net-nat rule. I still think we should consider updating the API (while the feature is Beta) and revisiting the current behavior, so that the NPL-allocated host port doesn't have to be the same for different protocols.

Reply:
Hi @antoninbas , I have offline synced up with AVI guys, and got a feedback that the solution that adding new annotation for a different Node port-Pod port-protocol mapping should not affect AVI's consumption on  the annotation, although they need to be aware of the change.

Reply:
PR updated. I implemented the first solution and now we support multiple annotations with the same nodeport.